internal/helm/repository.go

```diff
@@ -61,69 +61,85 @@ func NewChartRepository(repositoryURL string, providers getter.Providers, opts [
 }
 
 // Get returns the repo.ChartVersion for the given name, the version is expected
-// to be a semver.Constraints compatible string. If version is empty, the latest
-// stable version will be returned and prerelease versions will be ignored.
+// to be a semver.Constraints compatible string. In case of an ambiguous match,
+// takes the freshest chart. If version is empty, the latest stable version will
+// be returned and prerelease versions will be ignored.
 func (r *ChartRepository) Get(name, version string) (*repo.ChartVersion, error) {
-	cvs, ok := r.Index.Entries[name]
+	chartVersions, ok := r.Index.Entries[name]
 	if !ok {
 		return nil, repo.ErrNoChartName
 	}
-	if len(cvs) == 0 {
+	if len(chartVersions) == 0 {
 		return nil, repo.ErrNoChartVersion
 	}
 
-	// Check for exact matches first
-	if len(version) != 0 {
-		for _, cv := range cvs {
-			if version == cv.Version {
-				return cv, nil
-			}
-		}
-	}
-
-	// Continue to look for a (semantic) version match
-	latestStable := len(version) == 0 || version == "*"
-	var match semver.Range
-	if !latestStable {
-		rng, err := semver.ParseRange(version)
+	// Parse version constraints if given (default strategy: latest)
+	var versionLatestStable = len(version) == 0 || version == "*"
+	var versionConstraints *semver.Constraints
+	if !versionLatestStable {
+		rng, err := semver.NewConstraint(version)
 		if err != nil {
 			return nil, err
 		}
-		match = rng
+		versionConstraints = rng
 	}
-	var filteredVersions semver.Versions
-	lookup := make(map[string]*repo.ChartVersion)
-	for _, cv := range cvs {
-		v, err := semver.ParseTolerant(cv.Version)
+
+	// Filter out chart versions that doesn't satisfy constraints if any,
+	// parse semver and build a lookup table
+	matchingVersions := make(semver.Collection, 0)
+	chartVersionLookup := make(map[*semver.Version]*repo.ChartVersion)
+	for _, chartVersion := range chartVersions {
+		// Check for exact matches first
+		if version == chartVersion.Version {
+			return chartVersion, nil
+		}
```

Comment:
Checking for exact matches first is (without having benchmark proof) likely much faster, as `semver.NewVersion` / `semver.StrictNewVersion` do a lot of parsing.